File: spriteforest-html5/base/SFScene.js
;
/*!
 *
 *  SpriteForest JavaScript Library v1.0.0
 *  http://github.com/moky/SpriteForest/js/
 *
 *  Copyright (c) 2014 moKy at slanissue.com
 *  Released under the MIT license
 *  http://en.wikipedia.org/wiki/MIT_License
 *
 */
//
//  SFScene.js
//  SpriteForest-HTML5
//
//  Created by moKy on Feb. 26, 2014.
//

!function(sf) {
	'use strict';
	var cc = sf.cc;
	var cn = sf.cn;
	
	//
	//  constructor
	//
	var Scene = function() {
		cn.Scene.prototype.ctor.call(this);
	};
	
	//
	//  static functions
	//
	var create = function(dict) {
		if (arguments.length === 0 || !arguments[0]) {
			var node = new this();
			node.init();
			return node;
		} else if (typeof arguments[0] === 'string') {
			return sf.Node.create(arguments[0]);
		} else if (typeof arguments[0] === 'object') {
			var dict = arguments[0];
			var className = dict['Class'];
			while (!className) {
				dict['Class'] = 'SFScene';
				break;
			}
			return sf.Node.create(dict);
		} else {
			cn.error(arguments);
		}
		return null;
	};
	
	var addSpriteFramesWithFile = function(filename) {
		return cn.SpriteFrameCache.loadTextureAndSprites(filename);
	};
	
	var loadTextures = function(textures) {
		if (!textures || textures.length === 0) {
			return null;
		}
		var array = [];
		for (var i = 0; i < textures.length; ++i) {
			var filename = textures[i];
			if (typeof filename === 'object') {
				filename = filename['File'];
			}
			if (!filename) {
				cn.error(textures[i]);
				continue;
			}
			cn.log('texture filename = ' + filename);
			
			var texture = addSpriteFramesWithFile(filename);
			if (texture) {
				array.push(texture);
			} else {
				cn.log('failed to load texture: ' + filename);
			}
		}
		return array;
	};
	
	var cleanTextures = function(textures) {
		var frameCache = cc.SpriteFrameCache.getInstance();
		var textureCache = cc.TextureCache.getInstance();
		for (var i = textures.length - 1; i >= 0; --i) {
			var texture = textures[i];
			frameCache.removeSpriteFramesFromTexture(texture);
			TextureCache.removeTexture(texture);
		}
	};
	
	//
	//  instance functions
	//
	var initWithDictionary = function(dict) {
		if (!cn.Scene.prototype.init.call(this, dict)) {
			cn.error(this);
			return false;
		}
		
		// load textures
		var textures = dict['Textures'];
		loadTextures(textures);
		
		// add sub layers
		var layers = dict['Layers'];
		if (layers) {
			// scale top-level layers for N-Screen support
			for (var i = 0; i < layers.length; ++i) {
				var nodeDict = layers[i];
				if (nodeDict) {
					var scale = nodeDict['Scale'];
					var scaleX = nodeDict['ScaleX'];
					var scaleY = nodeDict['ScaleY'];
					if (scale === undefined && scaleX === undefined && scaleY === undefined) {
						nodeDict['Scale'] = 'ToFill';
					}
				}
			}
			this.addLayers(layers);
		}
		
		// add sub sprites
		//     WARNING: For simplify reason, I hope you should never use it!
		var sprites = dict['Sprites'];
		if (sprites) {
			this.addSprites(sprites);
		}
	};
	
	//--------------------------------------------------------------------------
	
	var prop = {
		ctor: Scene,
		
		initWithDictionary: initWithDictionary,
		
		// properties
	};
	
	// extends
	sf.Scene = cn.Scene.extend(prop);
	
	// static
	sf.Scene.create = create;
	
	sf.Scene.loadTextures = loadTextures;
	sf.Scene.cleanTextures = cleanTextures;
	
	// SFNode Extensions
	sf.Node.extensions(sf.Scene);
	
}(SpriteForest);

File: x/base/SFScene.cpp
//
//  SFScene.cpp
//  SpriteForest-X
//
//  Created by Moky on 12-11-29.
//  Copyright 2012 Slanissue Inc. All rights reserved.
//

#include "SFScene.h"

NS_SF_BEGIN

USING_NS_CC;
USING_NS_CN;

CCNode * SFScene::nodeWithDictionary(CCDictionary * pDict)
{
	CCString * pClassName = (CCString *)pDict->objectForKey("Class");
	if (!pClassName)
	{
		pDict->setObject(CCString::stringWithCString("SFScene"), "Class");
	}
	return SFNode::nodeWithDictionary(pDict);
}

bool SFScene::initWithDictionary(CCDictionary * pDict)
{
	if (!init())
	{
		CNLog("error");
		return false;
	}
	if (!pDict)
	{
		CNLog("error");
		return false;
	}
	
	
	// load textures
	CCArray * pTextures = CNDictionary::arrayForKey("Textures", pDict);
	SFScene::loadTextures(pTextures);
	
	// add sub layers
	CCArray * pLayers = CNDictionary::arrayForKey("Layers", pDict);
	if (pLayers)
	{
		// scale top-level layers for N-Screen support
		CCObject * pObj = NULL;
		CCARRAY_FOREACH(pLayers, pObj)
		{
			CCDictionary * pNodeDict = (CCDictionary *)pObj;
			if (pNodeDict)
			{
				CCString * pScale = (CCString *)pNodeDict->objectForKey("Scale");
				CCString * pScaleX = (CCString *)pNodeDict->objectForKey("ScaleX");
				CCString * pScaleY = (CCString *)pNodeDict->objectForKey("ScaleY");
				if (!pScale && !pScaleX && !pScaleY)
				{
					pNodeDict->setObject(CCString::stringWithCString("ToFill"), "Scale");
				}
			}
		}
	}
	addLayers(pLayers);
	
	// add sub sprites
	//     WARNING: For simplify reason, I hope you should never use it!
	CCArray * pSprites = CNDictionary::arrayForKey("Sprites", pDict);
	addSprites(pSprites);
	
	return true;
}

#pragma mark -

// copy from 'CCSpriteFrameCache.cpp'
CCTexture2D * SFScene_addSpriteFramesWithFile(const char * pszPlist)
{
	return coconut2d::CNSpriteFrameCache::loadTextureAndSprites(pszPlist); // high speed loader
//	
//    const char *pszPath = CCFileUtils::fullPathFromRelativePath(pszPlist);
//    CCDictionary *dict = CCDictionary::dictionaryWithContentsOfFileThreadSafe(pszPath);
//    
//	std::string texturePath("");
//	
//    CCDictionary* metadataDict = (CCDictionary*)dict->objectForKey("metadata");
//    if (metadataDict)
//    {
//        // try to read  texture file name from meta data
//        texturePath = metadataDict->valueForKey("textureFileName")->getCString();
//    }
//	
//    if (! texturePath.empty())
//    {
//        // build texture path relative to plist file
//        texturePath = CCFileUtils::fullPathFromRelativeFile(texturePath.c_str(), pszPath);
//    }
//    else
//    {
//        // build texture path by replacing file extension
//        texturePath = pszPath;
//		
//        // remove .xxx
//        size_t startPos = texturePath.find_last_of("."); 
//        texturePath = texturePath.erase(startPos);
//		
//        // append .png
//        texturePath = texturePath.append(".png");
//		
//        CNLog("Trying to use file %s as texture", texturePath.c_str());
//    }
//	
//    CCTexture2D *pTexture = CCTextureCache::sharedTextureCache()->addImage(texturePath.c_str());
//	
//    if (pTexture)
//    {
//		CCSpriteFrameCache::sharedSpriteFrameCache()->
//        addSpriteFramesWithDictionary(dict, pTexture);
//    }
//    else
//    {
//        CNLog("Couldn't load texture");
//    }
//	
//    dict->release();
//	
//	return pTexture;
}

CCArray * SFScene::loadTextures(CCArray * pTextures)
{
	if (!pTextures)
	{
		return NULL;
	}
	
	CCArray * pArray = CCArray::arrayWithCapacity(pTextures->count());
	
	CCPoolManager * pm = CCPoolManager::sharedPoolManager();
	
	CCObject * pObj = NULL;
	CCARRAY_FOREACH(pTextures, pObj)
	{
		pm->push();
		
		CCString * pFilename = NULL;
		
		if (CCDictionary * pDict = dynamic_cast<CCDictionary *>(pObj))
		{
			pFilename = (CCString *)pDict->objectForKey("File");
		}
		else if ((pFilename = dynamic_cast<CCString *>(pObj)))
		{
			std::string str(pFilename->getCString());
			str = CNFileManager::fullFilePath(str);
			pFilename = CCString::stringWithCString(str.c_str());
		}
		if (!pFilename)
		{
			pm->pop();
			continue;
		}
		CNLog("filename = %s", pFilename->getCString());
		
		CCTexture2D * pTexture = SFScene_addSpriteFramesWithFile(pFilename->getCString());
		if (pTexture)
		{
			pArray->addObject(pTexture);
		}
		else
		{
			CNLog("failed to load texture: %s", pFilename->getCString());
		}
		
		pm->pop();
	}
	
	return pArray;
}

void SFScene::cleanTextures(CCArray * pTextures)
{
	CCObject * pObj = NULL;
	CCARRAY_FOREACH(pTextures, pObj)
	{
		CCTexture2D * pTexture = dynamic_cast<CCTexture2D *>(pObj);
		if (!pTexture)
		{
			CNLog("error texture");
			continue;
		}
		CCSpriteFrameCache::sharedSpriteFrameCache()->removeSpriteFramesFromTexture(pTexture);
		CCTextureCache::sharedTextureCache()->removeTexture(pTexture);
	}
}

NS_SF_END
